src/cycle_basis.c

```diff
@@ -0,0 +1,660 @@
+    igraph_dqueue_t q=IGRAPH_DQUEUE_NULL;
+
+    long int i, j, k;
+
+    // Given a spanning tree (forest), fundamental cycles are given by
+    // unions of:
+    // - a path within the tree
+    // - a single edge connecting end to start nodes outside the tree
+    // The number of fundamental cycles is #edges - (#vertices -1),
+    // because the quantity in parenthesis is the number of edges in a
+    // tree (no cycles).
+    
+    // The key is that every non-tree edge generates a fundamental cycle
+    // so we run a BFS and whenever we encounter a vertex which is not
+    // new we are:
+    // - either trying to walk back the last edge we just walked
+    // - or using a non-tree edge (since both vertices are known already)
+    // Note: every edge must be walked once, so even after the tree
+    // is complete we still have to finish walking the remaining edges.
+    // Note: we cannot use the igraph_bfs function because we have to
+    // inspect outgoing edges that point to vertices that have already
+    // been added
+
+    IGRAPH_VECTOR_CHAR_INIT_FINALLY(&added_nodes, no_of_nodes);
+    IGRAPH_VECTOR_CHAR_INIT_FINALLY(&added_edges, no_of_edges);
+    IGRAPH_VECTOR_INIT_FINALLY(&parent_nodes, no_of_nodes);
+    IGRAPH_VECTOR_INIT_FINALLY(&parent_edges, no_of_edges);
+    IGRAPH_VECTOR_INIT_FINALLY(&dist, no_of_nodes);
+    IGRAPH_DQUEUE_INIT_FINALLY(&q, 100);
+    IGRAPH_VECTOR_INIT_FINALLY(&inc_edges, 0);
+
+    for (i = 0; i < no_of_nodes; i++) {
+        /* Start of a new connected component including the first one */
+        if (added_nodes[i] > 0) { continue; }
+
+        IGRAPH_ALLOW_INTERRUPTION();
+
+	/* This is the root, mark it visited */
+        added_nodes[i] = 1;
+
+	/* The queue stores the next vertex, the edge to it, and its distance to root */
+        IGRAPH_CHECK(igraph_dqueue_push(&q, i));
+        IGRAPH_CHECK(igraph_dqueue_push(&q, -1));
+        IGRAPH_CHECK(igraph_dqueue_push(&q, 0));
+        while (! igraph_dqueue_empty(&q)) {
+          /* current node, incoming edge, and distance from tree root */
+          long int actnode=(long int) igraph_dqueue_pop(&q);
+	  long int actedge=(long int) igraph_dqueue_pop(&q);
+          long int actdist=(long int) igraph_dqueue_pop(&q);
+
+	  /* check all edges connected to this vertex */
+          IGRAPH_CHECK(igraph_incident(graph, &inc_edges, (igraph_integer_t) actnode,
+				       IGRAPH_ALL));
+
+          for (j=0; j<igraph_vector_size(&inc_edges); j++) {
+              long int edge=(long int) VECTOR(inc_edges)[j];
+
+	      /* visited edges have been given their cycles already */
+              if (added_edges[edge] == 0) {
+
+		/* find the neighbour identity */
+                igraph_integer_t from, to;
+                igraph_edge(graph, (igraph_integer_t) edge, &from, &to);
+		
+		/* we only know that either one is actnode, swap if needed */
+                if (to==actnode) {
+		  to = from;
+		  from = to;
+		}
```

Comment:
Noting in case you were not aware:

We have `IGRAPH_TO` and `IGRAPH_FROM`, which are both simpler to use and more efficient than `igraph_edge` (they avoid any function call).

There is also `IGRAPH_OTHER`. If an undirected `edge` is `u-v` then `IGRAPH_OTHER(graph, edge, v)` gives `u`. Did you need `IGRAPH_OTHER(graph, edge, actnode)` here?

Reply:
We need to do this https://github.com/igraph/igraph/issues/1296

Reply:
This is also a copycat from `spanning_trees.c`. Shall I make a PR to fix that?

Reply:
I think it could be fixed in `spanning_trees.c` as well. That file is a very old part of the codebase, probably well before the time we've added `IGRAPH_TO` and `IGRAPH_FROM`. It probably does not make any difference performance-wise, but it helps people to pick up best practices if we have less places in the codebase where we don't use the shortcuts we've implemented to make our life easier in the long run.

Reply:
> It probably does not make any difference performance-wise

It depends. If this is *the only* function call in a loop, eliminating that call may very well make a noticeable difference. Function calls are expensive.